### ISSUE TITLE
Need -lopen-pal for lld when linking dynamically

### ISSUE THREAD

**Issue by ye-luo** [open-mpi/ompi#9059]
## Background information
AOMP 13.0-3 compiler uses lld as the default linker.
lld by default link with --no-allow-shlib-undefined
ld.lld: error: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so: undefined reference to opal_uses_threads [--no-allow-shlib-undefined]

To make OpenMPI compiler wrappers work for both ld and lld. My local fix is editing /usr/share/openmpi/mpiXXX-wrapper-data.txt
and adding `-lopen-pal` on the line of `libs` line
I'd like OpenMPI always add `-lopen-pal` for dynamic linking when installed instead of me editing wrapper data files by hand everywhere.

### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)
v4.0.3

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)
Ubuntu 20.04 apt-get installed

**Comment by jsquyres** [comment]
Can you try the latest 4.0.x release (either 4.0.5 or the recently-published 4.0.6rc4), or even the 4.1.1 release?

**Comment by ggouaillardet** [comment]
@ye-luo Please keep in mind the C++ bindings were removed from the MPI standard more than a decade ago, and they will likely be removed from future Open MPI versions as well.

**Comment by ye-luo** [comment]
If openmpi is configured with `--enable-mpi-cxx`
share/openmpi/mpicxx-wrapper-data.txt contains
```
libs=-lmpi_cxx -lmpi
```
and it went bad.

if openmpi has C++ binding disabled
share/openmpi/mpicxx-wrapper-data.txt contains
```
libs=-lmpi
```
works well.

I don't use C++ binding but I think for the ubuntu package maintainer, It is  also fair to enable C++ binding to support a wilder range of users. As long as OpenMPI is still shipping `--enable-mpi-cxx` for some users
It is better to write
```
libs=-lmpi_cxx -lopen-pal -lmpi
```
to avoid surprises to users

**Comment by ggouaillardet** [comment]
@ye-luo Fair enough, any given provided feature should work.
FWIW, I confirm the `--enable-mpi-cxx` option has been removed from the upcoming Open MPI v5 series.

I think the right fix here is not to update the wrapper, but have `libmpi_cxx.so` explicitly depends on `libopen-pal.so`.
Feel free to give #9060 a try - note , you will need recent autotools, and run `./autogen.pl` or `./autogen.pl --force` if you apply the patch to a tarball.

**Comment by ye-luo** [comment]
#9060 made lld happy in my test.

**Comment by ggouaillardet** [comment]
Great, thanks for the feedback!

**Comment by awlauria** [comment]
PR's merged, closing.

**Comment by baiyz** [comment]
Reproduced on latest openmpi 4.0.6, and verified by adding -lopen-pal after -lmpi_cxx

**Comment by ggouaillardet** [comment]
`4.0.6` was released before the fix was merged into the `v4.0.x` branch.

FWIW, the right fix is **not** to update the wrapper conf file, but to correctly link `libmpi_cxx.so`.
You can manually download and apply it from #9077 (note you will need recent autotools).
